### PR TITLE
Add AgnosticRawReader as a blocking wrapper around non-blocking text and binary readers

### PR DESCRIPTION
**Pull request by nirosys**
*Issue #, if available:* N/A

*Description of changes:*
This PR adds the format agnostic blocking raw reader, `BlockingRawReader`.

Changes:
* Existing blocking raw readers have been removed.
* A new trait has been introduced called `BufferedRawReader` which provides the functionality for adding data to a non-blocking raw reader, as well as marking a stream as complete. The end of stream marking was previously exposed via the RawReader trait.
* Type aliases have been added for `RawTextReader` and `RawBinaryReader` for referring to specific `BlockingRawReader` types, similarly to how the previous blocking implementations worked.
* Prior to this PR, the requirements around the non-blocking text reader forced all readers to implement End of Stream marking functionality. With this PR, those functions have been moved to the `BufferedRawReader` trait, and the non-blocking RawTextReader can now preemptively mark itself complete if data cannot be added to the buffer (currently just buffers of type `Vec<u8>` can add data).
*  There were a couple ion files that were added to the global skip list for the blocking binary reader. This should reflect the skip list for the non-blocking binary reader prior to this PR.
* A new function `raw_bytes` was added to the `BinaryBuffer` as a means to get access to the underlying (unshifted) bytes for use with the raw bytes API.
* The `RawBinaryBufferRawReader` was changed in order to force Incomplete errors through either `next` or `step_out` in order to match the behavior of the RawTextReader. This change included a new `ReaderState::WaitingForData(EncodedValue)` which waits for enough data to be added in order parse the entire encoded value. This stops incomplete errors from occurring on value materialization. This does not apply to container types.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/binary/non_blocking/raw_binary_reader.rs [hunk 1]
```diff
@@ -1343,10 +1543,10 @@ mod tests {
         expect_incomplete(reader.next());
         expect_incomplete(reader.next());
         // We can append data as it becomes available even if it doesn't produce a complete item.
-        reader.append_bytes(&[1, 0]);
+        reader.append_bytes(&[1, 0])?;
```
Comment (zslayton) [inline src/binary/non_blocking/raw_binary_reader.rs:1546]: Is `append_bytes` fallible now?

File: src/binary/non_blocking/raw_binary_reader.rs [hunk 2]
```diff
@@ -1788,14 +1990,102 @@ mod tests {
...
+        assert_eq!(cursor.field_id_offset(), Some(9));
+
+        cursor.step_out()?; // Step out of struct
+
+        // Second top-level value
+        assert_eq!(RawStreamItem::Value(IonType::Bool), cursor.next()?);
+        assert_eq!(cursor.raw_bytes(), Some(&ion_data[12..16]));
+        assert_eq!(cursor.raw_field_id_bytes(), None);
+        assert_eq!(cursor.raw_annotations_bytes(), Some(&ion_data[12..=14]));
+        // assert_eq!(cursor.annotations().next().unwrap()?, local_sid_token(12));
```
Comment (zslayton) [inline src/binary/non_blocking/raw_binary_reader.rs:2083]: Dangling comment?

File: src/text/non_blocking/raw_text_reader.rs [hunk 3]
```diff
@@ -96,33 +96,45 @@ pub(crate) enum RootParseResult<O> {
...
-impl<A: AsRef<[u8]>> RawTextReader<A> {
+impl From<Vec<u8>> for RawTextReader<Vec<u8>> {
+    fn from(source: Vec<u8>) -> Self {
+        RawTextReader::new(source)
+    }
+}
+
+pub trait Expandable {
+    fn expandable(&self) -> bool;
+}
```
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:107]: Very clever! Could you add a doc comment, please?

File: src/text/non_blocking/raw_text_reader.rs [hunk 4]
```diff
@@ -201,7 +213,7 @@ impl<A: AsRef<[u8]>> RawTextReader<A> {
                 // If the parser returns Ok(None), we've just encountered the end of the container for
                 // the first time. Set `is_exhausted` so we won't try to parse more until `step_out()` is
                 // called.
-                // We previously used a copy of the last `ParentLevel` in the stack to simplify reading.
+                // We previously used a copy of the last `ParentLevel` in the stack to smplify reading.
```
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:216]:
```suggestion
                // We previously used a copy of the last `ParentLevel` in the stack to simplify reading.
```

File: src/text/non_blocking/raw_text_reader.rs [hunk 5]
```diff
@@ -639,8 +651,9 @@ impl<A: AsRef<[u8]>> RawTextReader<A> {
     }
 }
 
-impl RawTextReader<Vec<u8>> {
-    pub fn append_bytes(&mut self, bytes: &[u8]) -> IonResult<()> {
+// impl RawTextReader<Vec<u8>> {
```
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:654]: Dangling comment?

File: src/text/non_blocking/raw_text_reader.rs [hunk 6]
```diff
@@ -1175,7 +1203,7 @@ mod reader_tests {
     #[case(" {{\"hello\"}} ", TextValue::Clob(Vec::from("hello".as_bytes())))]
     fn test_read_single_top_level_values(#[case] text: &str, #[case] expected_value: TextValue) {
         let reader = &mut RawTextReader::new(text);
-        reader.stream_complete();
+        // reader.stream_complete();
```
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:1206]: Dangling comments below?

Comment (nirosys) [reply src/text/non_blocking/raw_text_reader.rs:1206]: One day we will have the [technology](https://github.com/rust-lang/rust-clippy/issues/1348) .. >.>

File: src/binary/non_blocking/raw_binary_reader.rs [hunk 7]
```diff
@@ -365,6 +403,7 @@ impl<A: AsRef<[u8]>> RawBinaryBufferReader<A> {
             Skipping(bytes_to_skip) => bytes_to_skip,
             OnIvm => IVM.len(),
             OnValue(encoded_value) => encoded_value.total_length(),
+            WaitingForData(_) => unreachable!(),
```
[the remark this answers is not in the document]
Comment (nirosys) [reply src/binary/non_blocking/raw_binary_reader.rs:406]: Added in the next commit.

File: src/binary/non_blocking/raw_binary_reader.rs [hunk 8]
```diff
@@ -594,11 +774,31 @@ impl<A: AsRef<[u8]>> IonReader for RawBinaryBufferReader<A> {
         let mut tx_reader = self.transaction_reader();
 
         let item_result = tx_reader.read_next_item();
+        let nop_bytes_count = tx_reader.nop_bytes_count as usize;
+
+        // If we do not have enough bytes to materialize the next value, return an incomplete
+        // error. This is to match the behavior of the text reader where incomplets will only come
+        // from step-out and next calls.
+        if let ReaderState::OnValue(encoded_value) = tx_reader.state {
+            match encoded_value.ion_type() {
```
[the remark this answers is not in the document]
Comment (nirosys) [reply src/binary/non_blocking/raw_binary_reader.rs:783]: Switched over in the next commit, ty for pointing that out.

File: src/text/non_blocking/raw_text_reader.rs (hunk 3 above)
Comment (nirosys) [reply src/text/non_blocking/raw_text_reader.rs:107]: Absolutely, added in the next commit.

**Comment by nirosys** [comment]
Working on the conflicts now.

**Review by zslayton — APPROVED** [review]
:rocket:

**Pull request merged**